cubi_tk/irods_common.py

```diff
@@ -242,13 +261,115 @@ def get(self):
                     raise
                 except Exception as e:  # pragma: no cover
                     logger.error(f"Problem during transfer of {job.path_remote}")
                     logger.error(self.get_irods_error(e))
             t.clear()
+
+
+class iRODSRetrieveCollection(iRODSCommon):
+    """Class retrieves iRODS Collection associated with Assay"""
+
+    def __init__(
+        self, hash_scheme: str = DEFAULT_HASH_SCHEME, ask: bool = False, irods_env_path: Path = None
+    ):
+        """Constructor.
+
+        :param hash_scheme: iRODS hash scheme, default MD5.
+        :type hash_scheme: str, optional
+
+        :param ask: Confirm with user before certain actions.
+        :type ask: bool, optional
+
+        :param irods_env_path: Path to irods_environment.json
+        :type irods_env_path: pathlib.Path, optional
+        """
+        super().__init__(ask, irods_env_path)
+        self.hash_scheme = hash_scheme
+
+    def retrieve_irods_data_objects(self, irods_path: str) -> Dict[str, List[iRODSDataObject]]:
+        """Retrieve data objects from iRODS.
+
+        :param irods_path: iRODS path.
+
+        :return: Returns dictionary representation of iRODS collection information. Key: File name in iRODS (str);
+        Value: list of iRODSDataObject (native python-irodsclient object).
+        """
+
+        # Connect to iRODS
+        with self.session as session:
+            try:
+                root_coll = session.collections.get(irods_path)
+
+                # Get files and run checks
+                logger.info("Querying for data objects")
+
+                if root_coll is not None:
+                    irods_data_objs = self._irods_query(session, root_coll)
+                    irods_obj_dict = self.parse_irods_collection(irods_data_objs)
+                    return irods_obj_dict
+
+            except Exception as e:  # pragma: no cover
+                logger.error("Failed to retrieve iRODS path: %s", self.get_irods_error(e))
+                raise
+
+        return {}
+
+    def _irods_query(
+        self,
+        session: iRODSSession,
+        root_coll: iRODSCollection,
+    ) -> Dict[str, Union[Dict[str, iRODSDataObject], List[iRODSDataObject]]]:
+        """Get data objects recursively under the given iRODS path."""
+
+        ignore_schemes = [k.lower() for k in HASH_SCHEMES if k != self.hash_scheme.upper()]
+
+        query = session.query(DataObjectModel, CollectionModel).filter(
+            Like(CollectionModel.name, f"{root_coll.path}%")
+        )
+
+        data_objs = dict(files=[], checksums={})
+        for res in query:
+            # If the 'res' dict is not split into Colllection&Object the resulting iRODSDataObject is not fully functional,
+            # likely because a name/path/... attribute is overwritten somewhere
+            coll_res = {k: v for k, v in res.items() if k.icat_id >= 500}
+            obj_res = {k: v for k, v in res.items() if k.icat_id < 500}
```

Comment:
Perhaps store the `500` in some sensibly named constant, otherwise seems a bit arbitrary (at least to me, as someone who is not familiar with irods).

Reply:
500 seems to be the magic cutoff betwenn object & collection ids in irods. I could not really find documentation for that, but giving it a name sounds good.